Tasks/VsTest/vstest.ts

```diff
@@ -0,0 +1,515 @@
+                        defer.resolve(settingsFile);
+                    }
+                }
+            });
+        })
+            .fail(function(err) {
+                tl.debug("Error occured while overriding test run parameters. Continuing...");
+                tl.warning(err);
+                defer.resolve(settingsFile);
+            });
+    }
+    return defer.promise;
+}
+
+function isNugetRestoredAdapterPresent(rootDirectory: string): boolean {
+    var allFiles = tl.find(rootDirectory);
+    var adapterFiles = tl.match(allFiles, "**\\packages\\**\\*TestAdapter.dll", { matchBase: true });
+    if (adapterFiles && adapterFiles.length != 0) {
+        for (var i = 0; i < adapterFiles.length; i++) {
+            var adapterFile = adapterFiles[i];
+            var packageIndex = adapterFile.indexOf('packages') + 7;
+            var packageFolder = adapterFile.substr(0, packageIndex);
+            var parentFolder = path.dirname(packageFolder);
+            var solutionFiles = tl.match(allFiles, path.join(parentFolder, "*.sln"), { matchBase: true });
+            if (solutionFiles && solutionFiles.length != 0) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory: string): Q.Promise<string> {
+    var defer = Q.defer<string>();
+    if (!settingsFile || !pathExistsAsFile(settingsFile)) {
+        defer.resolve(defaultResultsDirectory);
+    }
+    else {
+        readFileContents(runSettingsFile, "utf-8").then(function(xmlContents) {
+            var parser = new xml2js.Parser();
+            parser.parseString(xmlContents, function(err, result) {
+                if (!err && result.RunSettings && result.RunSettings.RunConfiguration && result.RunSettings.RunConfiguration[0] &&
+                    result.RunSettings.RunConfiguration[0].ResultsDirectory && result.RunSettings.RunConfiguration[0].ResultsDirectory[0].length > 0) {
+                    defer.resolve(result.RunSettings.RunConfiguration[0].ResultsDirectory[0]);
+                }
+                else {
+                    defer.resolve(defaultResultsDirectory);
+                }
+            });
+        })
+            .fail(function(err) {
+                tl.debug("Error occured while reading test result directory from run settings. Continuing...")
+                tl.warning(err);
+                defer.resolve(defaultResultsDirectory);
+            });
+    }
+    return defer.promise;
+}
+
+function setupRunSettingsFileForParallel(settingsFile: string): Q.Promise<string> {
+    var defer = Q.defer<string>();
+    if (runInParallel) {
+        if (settingsFile && settingsFile.split('.').pop().toLowerCase() == "testsettings") {
+            tl.warning("Run in Parallel is not supported with testsettings file.");
+            defer.resolve(settingsFile);
+        }
+        else {
```

Comment:
avoid else blocks. return in case of error. success go ahead usually as discussed